rasa/core/nlg/generator.py

```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
             f"'{endpoint_config.type}'. Failed to create a "
             f"`NaturalLanguageGenerator` instance. Error: {e}"
         )
+
+
+class ResponseVariationFilter:
+    """Filters response variations based on the channel, action and condition."""
+
+    def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
+        self.responses = responses
+
+    @staticmethod
+    def _matches_filled_slots(
+        filled_slots: Dict[Text, Any], response: Dict[Text, Any]
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION, [])
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            filled_slots_value = filled_slots.get(name)
```

Comment:
I think its stylistically inconsistent to use `dict[key]` along with `dict.get(key)`. Can we use dict.get() in most cases?

Reply:
We always know that the constraints in a CRV will have those keys (name and value) because it's validated by the [YAML schema](https://github.com/RasaHQ/rasa/blob/6e10e42d052190577cafd9b22e79e367a60a5dc2/rasa/shared/nlu/training_data/schemas/responses.yml#L59), while filled_slots could have no value set for that slot, which is why it's needed to use `dict.get`. I also want to prevent adding checks for None to make mypy happy for those particular constraint keys. 

If you find it useful, I can add `required: True` to the keys in the yaml schema, this will definitely ensure that the keys are present.

Reply:
Sounds good, I don't think adding `required: True` is needed